src/era/rotating.py

```diff
@@ -0,0 +1,106 @@
+"""Create a new directory in base_dir and delete old directories as needed, logging as we go.
+"""
+
+import logging
+import re
+import shutil
+from datetime import datetime
+
+module_logger = logging.getLogger(__name__)
+
+
+class FolderRotator:  # pylint: disable=too-few-public-methods
+    """Create a new directory in base_dir and delete old directories as needed, logging as we go.
+    """
+
+    def __init__(self, base_dir) -> None:
+        self.base_dir = base_dir
+        self._class_logger = logging.getLogger(__name__).getChild(self.__class__.__name__)
+
+    #: Make dir Methods
+    def _build_new_download_dir_path(self, prefix, date_format):
+        today = datetime.today()
+        download_dir = self.base_dir / f'{prefix}{today.strftime(date_format)}'
+        return download_dir
+
+    def _make_new_download_dir(self, download_dir_path, exist_ok):
+        self._class_logger.debug(f'Attempting to create new directory `{download_dir_path}`')
+        try:
+            download_dir_path.mkdir(exist_ok=exist_ok)
+        except FileNotFoundError as error:
+            raise FileNotFoundError(f'Base directory `{self.base_dir}`` does not exist.') from error
+        else:
+            self._class_logger.debug(f'Successfully created `{download_dir_path}`')
+            return download_dir_path
+
+    #: Rotator Methods
+    def _get_all_but_n_most_recent_folder_paths(self, prefix, pattern, max_folder_count):
+        pattern = f'{prefix}{pattern}'
+        self._class_logger.debug(f'Folder regex pattern: {pattern}')
+        folder_paths = [path for path in self.base_dir.iterdir() if re.match(pattern, path.stem)]
+        if max_folder_count > len(folder_paths):
+            self._class_logger.debug(
+                f'max_folder_count `{max_folder_count}` greater than number of existing folders `{len(folder_paths)}`; '
+                'no folders deleted'
+            )
+            max_folder_count = 0
+        return sorted(folder_paths)[:-max_folder_count]
+
+    def _delete_old_folders(self, folder_paths_to_delete):
+        deleted_folders = []
+        for folder in folder_paths_to_delete:
+            try:
+                self._class_logger.debug(f'Attempting to delete `{folder}`')
+                shutil.rmtree(folder)
+            except Exception:
+                pass
```

Comment:
**question**: should the folders that can't be deleted get logged for manual intervention? Couldn't they foul up the next run when calling `_get_all_but_n_most_recent_folder_paths`

Reply:
I see what you mean... I put in the attempt log so that if you noticed it wasn't deleting folders as expected you could track it down with DEBUG level, but if each run introduced a new "sticky" folder eventually you'd get to the point where you're not deleting any folders and the directory could grow out of control. The script should create a new folder each time, so the only impact would be having too many files (I've seen very chatty logs [looking at you, samba] fill all available disk space and crash a server, but we're talking hundreds of lines of logging per second). 

Maybe it'd be best to log it as a warning or an error here instead of silently passing?

Reply:
I think a warning level log would be great here.